### ISSUE TITLE
IndexError: tuple index out of range

### ISSUE THREAD

**Issue by muliyangm** [ello0211/HLPT#5]
Hi, thanks for your great work! 

As I tried to deploy the code on my device, I ran into a bug in the following line using the given `finetune.sh`:

https://github.com/ello0211/HLPT/blob/a47a448d4282b7e07a5b423b9119235240f49c42/modeling_llama.py#L692

```
...
  File "/home/xxx/tools/anaconda3/envs/llm2/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/xxx/tools/anaconda3/envs/llm2/lib/python3.9/site-packages/transformers/models/llama/modeling_llama.py", line 692, in forward
    past_key_value = past_key_values[idx-24] if past_key_values is not None else None
IndexError: tuple index out of range
```

I think the issue may come from 
https://github.com/ello0211/HLPT/blob/a47a448d4282b7e07a5b423b9119235240f49c42/modeling_llama.py#L691
in which `23` may not be correct as the code printed (showing 32 layers in total):
```
...
base_model.model.layers.23.self_attn.prefix_gate.weight                                                                                                                                                                                 
base_model.model.layers.23.mlp.down_proj.lora_A.weight                                                                                                                                                                                  
base_model.model.layers.23.mlp.down_proj.lora_B.weight                                                                                                                                                                                  
base_model.model.layers.23.mlp.down_proj.lora_gate.weight                                                                                                                                                                               
base_model.model.layers.24.self_attn.prefix_gate.weight                                                                                                                                                                                 
base_model.model.layers.25.self_attn.prefix_gate.weight
base_model.model.layers.26.self_attn.prefix_gate.weight
base_model.model.layers.27.self_attn.prefix_gate.weight
base_model.model.layers.28.self_attn.prefix_gate.weight
base_model.model.layers.29.self_attn.prefix_gate.weight
base_model.model.layers.30.self_attn.prefix_gate.weight
base_model.model.layers.31.self_attn.prefix_gate.weight
prompt_encoder.embedding.weight
```
while in Line 692 `past_key_values` only has 3 elements and then throws out `IndexError: tuple index out of range`.

Could you please share your thoughts on this bug? Thanks!

**Comment by ello0211** [comment]
I'm really sorry for the inconvenience caused. This bug is mainly due an incorrect version uploaded by us. In line 204 of the /peft\src\peft\mapping.py code, which is used to set the number of prefix layers used at the top, the previous incorrect version was set to 3, which did not match with the modeling_llama.py. It has now been corrected to 8(Corresponding to the HLPT method), and there should be no further issues.

Thank you very much for your inquiry, as it helps us identify and fix the issues in the code. The current code is still relatively primitive, at an early stage of modification, There are still many areas in the code that can be optimized, and we will promptly work on improving them and provide updates in the readme. Thanks again for your question.

**Comment by muliyangm** [comment]
Thank you very much for your quick reply and effective help. Now I can successfully start training. By the way, there is another tiny issue I'd like to point out:

https://github.com/ello0211/HLPT/blob/e4b449d833f85aa11000b25422ee4f2962ed0ddb/peft/src/peft/tuners/lora.py#L181
This line will throw an error about initialization parameters as `key` will be mistakenly considered as the value to initialize `r` for the `Linear` class in Line 311. After modified into 
```
new_module = Linear(target.in_features, target.out_features, bias=bias, **kwargs)
```
then it can work as normal. Please let me know if this modification is problematic.

**Comment by muliyangm** [comment]
And, may I know what is the difference between HLPT and H2LPT?

**Comment by ello0211** [comment]
We believe this is not a big problem. and i will run my code again to make it clear as soon as i can.thank you

**Comment by ello0211** [comment]
Hi, have you successfully reproduced the code and results? We apologize for not replying promptly. We have carefully considered your questions. Firstly, regarding your first question, we apologize as this is still a problem caused by our version update. Your changes are correct and will not cause any issues. Regarding your second question, the main difference between HLPT and H2LPT lies in the number of layers. Currently, this code may require manual modification to halve the number of layers. For the application of lora in the bottom layer, you only need to change the number in lora.py of line 154 from 23 to 11. For the top layer prefix, you need to change the layer number in mapping.py from 8 to 4, and in modeling_llama.py, the code number should be changed to 27 and 28.

We must admit that the code is still relatively crude at the moment, and some parts require manual modification. However, we assure you that we will promptly optimize and correct the code. Thank you very much for your question.

**Comment by ello0211** [comment]
If you have any further questions, feel free to ask me anytime.

**Comment by ello0211** [comment]
Thank you very much for meticulously reproducing our results. We just carefully compared the result you obtained in evaluate.sh with the SVAMP dataset. It appears to be slightly lower by 3% compared to our tested result. This discrepancy may have been caused by some randomness in the search method during testing. However, it should not have a significant impact, as the results still demonstrate a relatively high accuracy.

**Comment by muliyangm** [comment]
I agree. Randomness can be unpredictable. Thanks again for your great work and kind reply! Wish you a good night (if it happens to be at night for you, too).

**Comment by ello0211** [comment]
Hi, happy new year! May I know do you get the final result (if it is convenient to tell me).

**Comment by muliyangm** [comment]
Hi, Happy Lunar New Year, too! 

Yes, I've tried several hyperparameter settings and can reach over 0.6 accuracy. The code itself sets up a good use case for combining prefix tuning and LoRA, yet I still have this academic question: Is prefix tuning + LoRA necessarily better than LoRA alone? Because I've also found that through some hyperparameter selections we can easily achieve 0.6+ accuracy by only using LoRA, too (using [LLM-Adapters](https://github.com/AGI-Edgerunners/LLM-Adapters)). One possible reason can be that the original LoRA (as in LLM-Adapters) fine-tunes $W_q$ and $W_v$ in all layers, yet in your hybrid implementation only the down projections in part of the layers are fine-tuned (I was also wondering the reason of choosing the down projection for LoRA tuning). Maybe in the end we still need exhaustive hyperparameter search to find the best choice combination.

**Comment by ello0211** [comment]
Hello, wishing you a Happy New Year again and hoping for your continued success in your work! By the way, you can also use `python multi_dataset_eval.py` to test the method on all datasets. The comprehensive average results across multiple datasets of the method may be more convincing. We are also willing to compare our comprehensive results with yours.
